core/src/main/java/com/linecorp/armeria/client/Http2ClientConnectionHandlerBuilder.java

```diff
@@ -36,6 +36,12 @@ final class Http2ClientConnectionHandlerBuilder
         this.protocol = protocol;
         // Disable RST frames limit for HTTP/2 clients.
         decoderEnforceMaxRstFramesPerWindow(0, 0);
+        final long gracefulShutdownTimeoutMillis = clientFactory.http2GracefulShutdownTimeoutMillis();
+        if (gracefulShutdownTimeoutMillis == Long.MAX_VALUE) {
+            gracefulShutdownTimeoutMillis(-1);
+        } else {
+            gracefulShutdownTimeoutMillis(gracefulShutdownTimeoutMillis);
+        }
```

Comment:
So is the current implementation of graceful shutdown in Armeria as follows? 
- Client: use netty's graceful shutdown feature
- Server: handled by Armeria

Reply:
Yes. Besides, we are going to add another `serverHttp2GracefultShutdownTimeoutMillis` that works after the drain duration after merging this PR. 😉

Reply:
Oh I see. Are the actions after server closes occur in below sequence? 
- server close -> connection drain(by Armeria) -> graceful shutdown(by Netty just like we did for the client ? 🤔 ) 

But isn't it a bit weird to execute graceful shutdown after connection draining? Are there any more connections left to gracefully shutdown?

Reply:
The purpose of connection drain and graceful shutdown is different.
- During connection drain, a server receives inflight(new) requests.
- During graceful shutdown, a server refuses new requests but waits for requests being processed to be finished.

The process for connection shutdown on the server side will follow:
- A server sends `GO_AWAY(Integer.MAX_VALUE)` and start connection drain.
https://github.com/line/armeria/blob/a586c3d24874125f3e0a26141581407fc1f18b4e/core/src/main/java/com/linecorp/armeria/server/Http2ServerConnectionHandler.java#L175-L175
- A client receives `GO_AWAY(Integer.MAX_VALUE)` and makes the connection inactive. But it does not close the connection.
https://github.com/line/armeria/blob/a586c3d24874125f3e0a26141581407fc1f18b4e/core/src/main/java/com/linecorp/armeria/client/Http2ResponseDecoder.java#L168-L169
https://github.com/line/armeria/blob/a586c3d24874125f3e0a26141581407fc1f18b4e/core/src/main/java/com/linecorp/armeria/internal/common/Http2GoAwayHandler.java#L80-L86
- The server waits 1 second to finish the connection drain and then sends 
  `GO_AWAY(last_stream_id)` by Netty.
- The server refuses new requests.
- A graceful shutdown is performed for up to 30 seconds. Without an active stream,  the graceful shutdown may be finished early.

Reply:
What a crystal clear explanation. Thank you so much for the detail. 🙇